### PR TITLE
Upgrade to AWS SDK v2

### PR DESCRIPTION
**Pull request by mngo87**
*Issue #, if available:*
https://github.com/awslabs/amazon-sns-java-extended-client-lib/issues/7

*Description of changes:*
Ensure this library is compatible with Java SDK V2

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/software/amazon/sns/AmazonSNSExtendedClient.java [hunk 1]
```diff
@@ -79,28 +82,30 @@ public AmazonSNSExtendedClient(AmazonSNS snsClient, SNSExtendedClientConfigurati
...
+        if (!StringUtils.isNullOrEmpty(publishRequest.messageStructure()) &&
+                publishRequest.messageStructure().equals(MULTIPLE_PROTOCOL_MESSAGE_STRUCTURE)) {
             String errorMessage = "SNS extended client does not support sending JSON messages.";
             LOGGER.error(errorMessage);
             throw new AmazonClientException(errorMessage);
         }
 
-        publishRequest.getRequestClientOptions().appendUserAgent(USER_AGENT_HEADER);
+        PublishRequest.Builder publishRequestBuilder = publishRequest.toBuilder();
+        publishRequestBuilder.overrideConfiguration(AwsRequestOverrideConfiguration.builder().putHeader(USER_AGENT_HEADER_NAME, USER_AGENT_HEADER).build());
```
Comment (mngo87) [inline src/main/java/software/amazon/sns/AmazonSNSExtendedClient.java:98]: I wasn't sure if there was a better way to handle this. Please advise here

**Comment by mngo87** [comment]
I am unable to specify reviewers. so tagging @aws-ivvladim @aws-rizi

File: src/main/java/software/amazon/sns/AmazonSNSExtendedClient.java [hunk 2]
```diff
@@ -79,28 +82,30 @@ public AmazonSNSExtendedClient(AmazonSNS snsClient, SNSExtendedClientConfigurati
      * Documentation</a>
```
Comment (aws-rizi) [inline src/main/java/software/amazon/sns/AmazonSNSExtendedClient.java:82]: Please update all the Java docs in this file and AmazonSNSExtendedClientBase.java to match java docs in SDK V2.

File: src/main/java/software/amazon/sns/AmazonSNSExtendedClientBase.java [hunk 3]
```diff
@@ -1,21 +1,13 @@
 package software.amazon.sns;
```
Comment (aws-rizi) [inline src/main/java/software/amazon/sns/AmazonSNSExtendedClientBase.java:1]: As mentioned on  all the java docs should be updated to fully match current java docs in SDK V2.

File: src/main/java/software/amazon/sns/AmazonSNSExtendedClientBase.java [hunk 4]
```diff
@@ -32,69 +24,17 @@ public AmazonSNSExtendedClientBase(AmazonSNS amazonSNSToBeExtended) {
      * @throws InternalErrorException      Indicates an internal service error.
      * @throws AuthorizationErrorException Indicates that the user has been denied access to the requested resource.
      * @throws NotFoundException           Indicates that the requested resource does not exist.
+     * @throws SdkClientException          If any client side error occurs such as an IO related failure, failure to get credentials, etc.
+     * @throws SnsException                Base class for all service exceptions. Unknown exceptions will be thrown as an instance of this type.
      * @sample AmazonSNS.GetEndpointAttributes
```
Comment (aws-rizi) [inline src/main/java/software/amazon/sns/AmazonSNSExtendedClientBase.java:29]: These lines should be ```@sample SnsClinet.```

**Review by aws-rizi — CHANGES_REQUESTED** [review]

**Comment by mngo87** [comment]
Thanks for the review @aws-rizi. I updated the comments per your request

File: pom.xml [hunk 5]
```diff
@@ -6,7 +6,7 @@
 
     <groupId>software.amazon.sns</groupId>
     <artifactId>sns-extended-client</artifactId>
-    <version>1.0.0</version>
+    <version>2.0.0</version>
```
Comment (aws-rizi) [inline pom.xml:9]: I believe new version of payload-offloading-java-common-lib-for-aws is published. Can you please update the file to use the new version on that library?

```
<dependency>
            <groupId>software.amazon.payloadoffloading</groupId>
            <artifactId>payloadoffloading-common</artifactId>
            <version>1.0.0</version>.  ===> should be changed to <version>2.0.0</version>
</dependency>
```

**Review by aws-rizi — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 6]
```diff
@@ -44,7 +44,7 @@
         <dependency>
             <groupId>software.amazon.payloadoffloading</groupId>
             <artifactId>payloadoffloading-common</artifactId>
-            <version>1.0.0</version>
+            <version>2.0.0</version>
         </dependency>
         <dependency>
             <groupId>com.amazonaws</groupId>
```
Comment (mngo87) [inline pom.xml:50]: note that until `amazon-sqs-java-extended-client-lib` is updated to use aws sdk v2, there will still be some references to aws sdk v1 in the code. i put up a PR to update the sqs sdk as well (https://github.com/awslabs/amazon-sqs-java-extended-client-lib/pull/58)

**Comment by markmansur** [comment]
Hey any updates on this? Support for SDK V2 would be awesome!

**Comment by stringbean** [comment]
Now that https://github.com/awslabs/amazon-sqs-java-extended-client-lib/pull/58 has been merged can we get the dependency updated in this PR and get it merged in?

File: pom.xml (hunk 6 above)
Comment (agummi-amazon) [reply pom.xml:50]: updated to `amazon-sqs-java-extended-client-lib` v2.0.2

File: pom.xml (hunk 5 above)
Comment (agummi-amazon) [reply pom.xml:9]: updated the payload-offloading-java-common-lib-for-aws version to v 2.1.2

File: src/main/java/software/amazon/sns/AmazonSNSExtendedClient.java [hunk 7]
```diff
@@ -256,6 +256,7 @@ private PublishRequest storeMessageInExtendedStore(PublishRequest publishRequest
         messageAttributeValueBuilder.stringValue(messageContentSize.toString());
         MessageAttributeValue messageAttributeValue = messageAttributeValueBuilder.build();
 
+
```
Comment (shih) [inline src/main/java/software/amazon/sns/AmazonSNSExtendedClient.java:259]: nit: Unnecessary line here.

**Review by shih — APPROVED** [review]

**Pull request merged**